refactor: merge all syncs in one place (AR-1438)

vitorhugods opened the pull request:
----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [X] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [X] contains a reference JIRA issue number like `SQPIT-764`
  - [X] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [X] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

- The different stages of Sync are being done independently from one another and there's no single source of thruth, that all the steps happen in order.
- There's no guarantee that we are not processing the same events twice.
- Exceptions leak when using `ListenToEventsUseCase`.

### Solutions

- Move events processing together with Slow sync, and tie them together so they happen in order
- Limited parallelism for sync events processing, so we can't have two syncs happening at the same time.
- Replace `COMPLETE` with `LIVE` regarding Sync status. 
- Add `PROCESSING_LIVE_EVENTS` Sync status.
- Rename `waitForSlowSyncToComplete` to `waitForSyncToComplete`, as all steps are together now.


### Testing

Extensively manually tested using Reloaded. 

Automated tests are under development.

#### Test Coverage

- [X] I have added automated test to this contribution

#### How to Test

See the attachment below:

### Attachments

See [this video](https://youtu.be/xnyg9fNzyTs) for a demo.

### Notes

I'm thinking about in next PRs:
- Replace most of the utilizations of `waitForSyncToComplete` with a `startSync`, that doesn't block the caller, as most UseCases don't quite require sync to be complete and data can be observed before sync is done.
- Create a platform-specific NetworkListener that will monitor network status and start sync once it's connected (eager sync)
- Create a ObserveSyncStatusUseCase so clients can display a progress bar and offline indications.

----
#### PR Post Merge Checklist for internal contributors

 - [X] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

vitorhugods commented on logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt:
```diff
@@ -265,9 +281,9 @@ abstract class UserSessionScopeCommon(
     private val logoutRepository: LogoutRepository = LogoutDataSource(authenticatedDataSourceSet.authenticatedNetworkContainer.logoutApi)
     val listenToEvents: ListenToEventsUseCase
-        get() = ListenToEventsUseCase(syncManager, eventRepository, conversationEventReceiver)
+        get() = ListenToEventsUseCase(syncManager)
```
This can be removed in a next PR.

Other UseCases will trigger sync and "listening to events".

trOnk12 reviewed (commented):
aint gonna approve it as I do not have that expertise :P but, great job, clean, safe,  well thought

MohamadJaara commented on logic/src/commonMain/kotlin/com/wire/kalium/logic/sync/SyncManager.kt:
```diff
@@ -1,43 +1,176 @@
 class SyncManagerImpl(
     private val workScheduler: WorkScheduler,
-    private val syncRepository: SyncRepository
+    private val eventRepository: EventRepository,
+    kaliumDispatcher: KaliumDispatcher,
+    private val syncRepository: SyncRepository,
+    private val conversationEventReceiver: ConversationEventReceiver
 ) : SyncManager {
```
```suggestion
    private val eventRepository: EventRepository,
    private val syncRepository: SyncRepository,
    private val conversationEventReceiver: ConversationEventReceiver
        kaliumDispatcher: KaliumDispatcher = KaliumDispatcherImpl,
) : SyncManager {
```
KaliumDispatcher can be injected into constructor

MohamadJaara commented on logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/UserSessionScope.kt:
```diff
@@ -211,7 +216,17 @@ abstract class UserSessionScopeCommon(
+        SyncManagerImpl(
+            authenticatedDataSourceSet.workScheduler,
+            eventRepository,
+            KaliumDispatcherImpl,
```
```suggestion
```
can be removed if KaliumDispatcherImpl is injected into constructor

MohamadJaara reviewed (approved):
Excellent job 🚀, only one tiny comment about injecting KaliumDispatcher into constructor

vitorhugods commented:
I deleted a coment of yours trying to delete mine 😓. Completely on accident. Was the one about changing `SyncManagerImpl`'s constructor.

I've addressed the changes.
